File: Litvinov group/semester 1/06.HW/6.2/stack.cpp
#include "stack.h"

#include <iostream>
#include <string>

using namespace std;

struct StackElement {
    char value;
    StackElement *next;
};

struct Stack {
    StackElement *head;
};

Stack* createStack() {
    return new Stack{};
}

void push(Stack *stack, char value) {
    StackElement *newElement = new StackElement{value, stack->head};
    stack->head = newElement;
}

char pop(Stack *stack) {
    StackElement *temp = stack->head;
    stack->head = temp->next;
    char answer = temp->value;
    delete temp;
    return answer;
}

bool isEmpty(Stack* stack) {
    return stack->head == nullptr;
}

void deleteStack(Stack* stack) {
    while(!isEmpty(stack)) {
        int a = pop(stack);
    }
    delete stack;
}

StackElement* head(Stack* stack){
    return stack->head;
}

File: Litvinov group/semester 2/2.HW/2.4/2.4/Stack.cs
using static System.Console;

namespace StackCalculator
{
    /// <summary>
    /// Stack on pointers
    /// </summary>
    public class Stack : IStack
    {
        /// <summary>
        /// The ELement for stack
        /// </summary>
        private class Element
        {
            public Element(string str, Element next)
            {
                this.next = next;
                this.str = str;
            }

            public Element next;
            public string str;
        }

        /// <summary>
        /// Push the element into stack
        /// </summary>
        /// <param name="value">Element to be added</param>
        public void Push(string value)
        {
            top = new Element(value, top);
            ++size;
        }

        /// <summary>
        /// Get the last value and delete it from stack
        /// </summary>
        /// <returns>Wanted value</returns>
        public string Pop()
        {
            if (top == null)
            {
                WriteLine("Stack`s empty");
                return "";
            }

            Element element = top;
            top = top.next;
            --size;

            return element.str;
        }

        /// <summary>
        /// Check whether the stack`s empty
        /// </summary>
        /// <returns>Answer</returns>
        public bool IsEmpty() => top == null;

        /// <summary>
        /// get size of the stack
        /// </summary>
        /// <returns>Wanted size</returns>
        public int Size() => size;

        private Element top;
        private int size;
    }
}
